### DISCUSSION TITLE
Schedules are slow to evaluate

### DISCUSSION THREAD

**Discussion by nsteins** [dagster-io/dagster#23933] [Q&A] [answered]
As we've scaled up the number of assets and schedules in our dagster deployment, we've found that the evaluation time for schedule ticks can be very long, which is a problem since we are trying to run some jobs once every 5 minutes. At their best, schedule ticks are taking ~40 seconds to evaluate, but when there are many concurrent ticks (which we see at the top of the hour), it can take 10-15 minutes to evaluate a single schedule tick.

We've tried increasing the number of workers for the schedules, increasing cpu request limits for the daemon-server and upgrading the db instance to increase the # of cpus. Any advice on where to optimize or where to search for bottlenecks?

**Comment by gibsondan** [comment]
Hi @nsteins - running py-spy (https://github.com/dagster-io/dagster/discussions/14771) on your daemon and your code server while schedules are running slowly could help give some insight into the source of the bottleneck here. If it's possible to include a speedscope output here from those processes while things are running slowly, we could take a closer look.

Every schedule tick needs to run code on your code server task, so increasing the amount of CPU available to that process/task/container would be one possibility to consider.

**Reply by nsteins** [reply]
Hi @gibsondan, sorry for the late response - we've been iterating on this issue and tried several solutions:

- spreading out the ticks to limit concurrent job schedules. This has helped most, as we have a large number of assets that execute between hourly and daily, so we have much more freedom to spread them out over a wider period of time.
- we've increased `numScheduleWorkers` which seems to have limited effect
- we've increased the db to an 8 cpu instance, which is more expensive than we would like, but we no longer see db cpu pegged at 90-100%
- we are purging old db data to keep the tables as small as possible while retaining recent data

However, we're now seeing that schedule evaluation can sometimes still be slow, though not consistently. There are periods of schedule slowness that do not have an obvious cause. Both the daemon and code server instances report using 500-1500 mcpu out of the 4000mcpu they have requested, so I do not think they are resource constrained there. Our schedule tick code is fairly simple - generates a list of dates (in the current case 2) and yields RunRequest for each date. We have around 150 assets with schedules that run 2 materializations every 5 minutes, each with sensors that run another 2 materialization per partition when those jobs are complete.

I'm attaching pyspy speedscope records from a time when schedule evaluation was particularly slow. These were not taken concurrently, and I had to reduce the sampling rate to 5 to avoid it running behind schedule. Let me know if there is any other information that would be helpful to share.

[speedscope-code-server-grpc.json](https://github.com/user-attachments/files/17078709/speedscope-code-server-grpc.json)
[speedscope-code-server.json](https://github.com/user-attachments/files/17078710/speedscope-code-server.json)
[speedscope-daemon.json](https://github.com/user-attachments/files/17078711/speedscope-daemon.json)

**Reply by gibsondan** [reply]
Hi @nsteins - looking at the daemon speedscope, it looks like your daemon is spending a lot of time just doing json deserialization of the information that it pulls from the database. What's the current CPU available for the daemon pod?

**Reply by gibsondan** [reply]
Also how many schedules do you have in total?

**Reply by gibsondan** [reply]
Ah you said that earlier. The speedscope that you showed earlier shows 13 seconds just deserializing the rows from the database that say which schedules are running, which feels unusually slow: 
<img width="1077" alt="image" src="https://github.com/user-attachments/assets/170a2626-6edd-489b-bfa5-997b5011e6ba">

**Reply by gibsondan** [reply]
The other command that can be useful when lots of different threads are in the mix is `py-spy top` which will tell you what is taking up the most CPU over a given period of time.

**Reply by gibsondan** [reply]
Can you confirm that you generated the speedscope with the "--idle" flag? I'm seeing the threads on the daemon have very different time lengths  - the idle flag will let you know how they spent their time if they were running but idle (for example, waiting for a database connection)

**Reply by gibsondan** [reply]
And one last thing - can you try running this command in your db and tell us what count it returns:
```
SELECT COUNT(1) FROM jobs WHERE job_type='SCHEDULE';
```

**Reply by nsteins** [reply]
Thanks so much for your response, @gibsondan!

- The daemon and the code server both have 4000 mcpu available through request, and no limit set, so it's possible they could use more. However, I have never observed them using more than 1500mcpu
- We have a total of 573 schedules on this code server, plus 553 running on different code server, but a large number of those run either daily or hourly and are spread out to not run concurrently. The number running every 5 minutes is ~150, and that is where we have little room to spread out the execution times.
- I did not use the `--idle` flag in the previous speedscopes, I'm uploading new results with the idle flag run.

[speedscope-server-grpc-idle.json](https://github.com/user-attachments/files/17079739/speedscope-server-grpc-idle.json)
[speedscope-server-idle.json](https://github.com/user-attachments/files/17079741/speedscope-server-idle.json)
[speedscope-server-daemon-idle.json](https://github.com/user-attachments/files/17080111/speedscope-server-daemon-idle.json)
[speedscope-server-grpc-idle.json](https://github.com/user-attachments/files/17080113/speedscope-server-grpc-idle.json)
[Uploading speedscope-server-idle.json…]()


**py-spy top on the daemon-server**
Total Samples 5500, Error Rate 3.47% (No such file or directory (os error 2))
GIL: 45.00%, Active: 220.00%, Threads: 105

  %Own   %Total  OwnTime  TotalTime  Function (filename)
 88.00%  88.00%   107.6s    107.6s   _acquireLock (logging/__init__.py)
 22.00%  22.00%   15.07s    15.20s   connect (psycopg2/__init__.py)
 22.00%  51.00%   10.27s    17.52s   unpack (dagster/_serdes/serdes.py)
 16.00%  16.00%    9.18s     9.18s   do_execute (sqlalchemy/engine/default.py)
  6.00%  96.00%    5.95s    122.2s   _worker (concurrent/futures/thread.py)
  8.00%  60.00%    5.03s    22.97s   _unpack_object (dagster/_serdes/serdes.py)
  6.00%   6.00%    4.37s     4.39s   _blocking (grpc/_channel.py)
 10.00%  70.00%    4.01s    26.98s   raw_decode (json/decoder.py)
  5.00%   5.00%    3.04s     3.04s   do_close (sqlalchemy/engine/default.py)
  6.00%   6.00%    1.16s     1.16s   <lambda> (<string>)
  0.00%   0.00%   0.880s     2.32s   iterencode (json/encoder.py)
  7.00%  13.00%   0.720s     1.92s   __checked_new__ (<string>)
  2.00%  75.00%   0.710s    31.94s   deserialize_values (dagster/_serdes/serdes.py)
  0.00%   0.00%   0.690s    0.690s   set_isolation_level (sqlalchemy/dialects/postgresql/psycopg2.py)
  0.00%   0.00%   0.670s     1.21s   pack_items (dagster/_serdes/serdes.py)
  0.00%   0.00%   0.570s    0.570s   _next (grpc/_channel.py)
  0.00%   0.00%   0.540s    0.770s   _transform_for_serialization (dagster/_serdes/serdes.py)
  2.00%   5.00%   0.540s     1.03s   __new__ (dagster/_core/scheduler/instigation.py)
  0.00%  70.00%   0.510s    27.49s   decode (json/decoder.py)
  0.00%   1.00%   0.490s     1.15s   __new__ (dagster/_core/remote_representation/origin.py)
  0.00%   0.00%   0.470s    0.470s   __init__ (json/decoder.py)
  2.00%   2.00%   0.470s     1.64s   disable_dagster_warnings (dagster/_utils/warnings.py)
  0.00%   0.00%   0.460s     1.72s   __new__ (dagster/_config/snap.py)
  0.00%  56.00%   0.430s    16.51s   check_daemon_loop (dagster/_daemon/controller.py)
  0.00%   0.00%   0.430s    0.510s   channel_spin (grpc/_channel.py)
  1.00%  71.00%   0.410s    28.37s   loads (json/__init__.py)
  2.00%   2.00%   0.400s    0.400s   __init__ (contextlib.py)
  0.00%   0.00%   0.380s    0.380s   hash_str (dagster/_serdes/utils.py)
  0.00%   0.00%   0.370s    0.380s   do_rollback (sqlalchemy/engine/default.py)
  1.00%   1.00%   0.350s    0.350s   str_param (dagster/_check/functions.py)
  0.00%   0.00%   0.340s    0.340s   __getitem__ (enum.py)
  0.00%  19.00%   0.320s    20.44s   __exit__ (contextlib.py)

**py-spy top on the code-server**

  %Own   %Total  OwnTime  TotalTime  Function (filename)
1168.00% 1200.00%   571.3s    576.0s   _worker (concurrent/futures/thread.py)
500.00% 500.00%   239.8s    239.8s   wait (threading.py)
100.00% 100.00%   47.96s    48.00s   _serve (grpc/_server.py)
 32.00%  32.00%    4.44s     4.47s   _blocking (grpc/_channel.py)
  0.00%   0.00%   0.060s    0.060s   _close (grpc/_channel.py)
  0.00%   0.00%   0.060s    0.070s   _status (grpc/_server.py)
  0.00%   0.00%   0.050s    0.110s   _channel_managed_call_management (grpc/_channel.py)
  0.00%   0.00%   0.030s    0.030s   __hash__ (typing.py)
  0.00%   0.00%   0.020s    0.020s   encode (grpc/_common.py)
  0.00%   0.00%   0.020s    0.150s   __init__ (dagster/_grpc/__generated__/api_pb2_grpc.py)
  0.00%   0.00%   0.020s    0.020s   __init__ (grpc/_channel.py)
  0.00%   0.00%   0.020s    0.020s   __exit__ (threading.py)
  0.00%   0.00%   0.020s    0.060s   inner (typing.py)
  0.00%   0.00%   0.020s    0.070s   unary_request (grpc/_server.py)
  0.00%   0.00%   0.010s    0.010s   notify_all (threading.py)
  0.00%   0.00%   0.010s    0.070s   __exit__ (contextlib.py)
  0.00%   0.00%   0.010s    0.010s   _is_owned (threading.py)
  0.00%   0.00%   0.010s    0.010s   _handle_event (grpc/_channel.py)
  0.00%   0.00%   0.010s    0.010s   receive_close_on_server (grpc/_server.py)
  0.00%   0.00%   0.010s    0.010s   __eq__ (typing.py)
  0.00%   0.00%   0.010s    0.210s   _query (dagster/_grpc/client.py)
  0.00%   0.00%   0.010s    0.010s   set_result (concurrent/futures/_base.py)
  0.00%   0.00%   0.010s    0.010s   send_status_from_server (grpc/_server.py)
  0.00%   0.00%   0.010s    0.010s   __enter__ (threading.py)
  0.00%   0.00%   0.010s    0.010s   submit (concurrent/futures/thread.py)
  0.00%  32.00%   0.000s     4.66s   run (concurrent/futures/thread.py)
  0.00%   0.00%   0.000s    0.020s   __enter__ (contextlib.py)
  0.00% 1700.00%   0.000s    816.0s   _bootstrap_inner (threading.py)
  0.00% 100.00%   0.000s    48.00s   _clear_old_processes (dagster/_core/remote_representation/grpc_server_registry.py)
  0.00% 1700.00%   0.000s    816.0s   run (threading.py)
  0.00% 100.00%   0.000s    48.00s   __call__ (click/core.py)
  0.00%   0.00%   0.000s    0.010s   _handle_call (grpc/_server.py)
  0.00% 100.00%   0.000s    48.00s   server_termination_target (dagster/_grpc/server.py)
  0.00% 100.00%   0.000s    48.00s   _wait_once (grpc/_common.py)
  0.00% 100.00%   0.000s    48.00s   main (click/core.py)
  0.00%   0.00%   0.000s    0.060s   ExecutionPlanSnapshot (dagster/_g

**Reply by nsteins** [reply]
And running SELECT COUNT(1) FROM jobs WHERE job_type='SCHEDULE'; yields `1225`

**Reply by gibsondan** [reply]
```
88.00% 88.00% 107.6s 107.6s _acquireLock (logging/init.py)
```

This is really interesting and suggests that the majority of the time is actually being spent contending over the global python logging module's lock.

**Reply by nsteins** [reply]
Is that possibly related to the large number of threads? we have increased the number of workers for the scheduler, sensor, and run queue

**Reply by gibsondan** [reply]
This is not at all where I thought this thread would go, but this appears to be at least part of what is going on here (just filed this issue): https://github.com/kubernetes-client/python/issues/2284

**Reply by gibsondan** [reply, marked as answer]
And https://github.com/dagster-io/dagster/pull/24635 fixes something similar in our own code that was modeled after the k8s api client's model deserialization that I can also see appearing in those speedscopes.

I do think you could try running with fewer threads and see if that improves contention here. You could also see if things get better if you try running the daemon at a higher log level to reduce logger contention (There is a --log-level argument to the "dagster-daemon run" command, although I think we may not expose it directly in the helm chart)

**Reply by nsteins** [reply]
Thank you very much for your help in diagnosing this!

I am currently trying disabling the schedule submit threadpool as I noticed in the py-spy logs that the schedule daemon worker threads were mostly waiting, while the submit worker threads were the ones submitting run requests, which is the time-limiting process due to the acquire locks. I will continue to test different configurations with the worker threads, but this appears to be helping things already.

![Screenshot 2024-09-20 at 4 32 20 PM](https://github.com/user-attachments/assets/ac963dc3-c2a4-4942-a2e5-7c21241e9495)
![Screenshot 2024-09-20 at 4 32 07 PM](https://github.com/user-attachments/assets/ca84c2b6-c4ac-403a-bf2d-5996249b7cc0)

**Reply by gibsondan** [reply]
sounds good, thanks for reporting this. It seems like this is a longstanding known issue with the k8s python client (it's possible there were other sources of contention, but this certainly jumps out as a likely culprit). We're evaluating the pros and cons of patching the client to work around it.

**Reply by CPAPI-104** [reply]
@gibsondan Any ETA on when the fix you mentioned [here](https://github.com/dagster-io/dagster/pull/24635) will be merged and released?

Our sensors should be running every 60 seconds, but it's taking 6.5 minutes

**Reply by gibsondan** [reply]
I think we can include it in our open source release next week. You may want to verify with profiling that that is actually the cause of your issue, I would only expect it to be affecting you if you have many concurrent threads launching runs over the kubernetes api in your daemon.

**Reply by CPAPI-104** [reply]
@gibsondan Hey wanted to follow up to see if this was included in the latest release? If it is it'd be worth it for us to do a major version upgrade for us

We'll also be taking some time soon to give a good thorough check on performance in the cluster with some py-spy logs included. So I'll make sure to include those here when we get them.

**Reply by gibsondan** [reply]
Sorry for not following up here. This went out in dagster 1.8.10 last week.
